### PR TITLE
[TT-13359] move upstream basic auth to ee package

### PR DESCRIPTION
**Pull request by jeffy-mathew**
## Description

This PR moves upstream basic auth implementations to ee package
## Related Issue
Parent: https://tyktech.atlassian.net/browse/TT-13359
Subtask: https://tyktech.atlassian.net/browse/TT-13389

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

___

### **PR Type**
Enhancement, Other

___

### **Description**
- Refactored `UpstreamBasicAuth` to use a new `AuthSource` struct for better configuration management.
- Implemented a new middleware for upstream basic authentication, including lifecycle and request processing methods.
- Integrated the new middleware into the API processing chain.
- Added EE-specific implementations and build constraints to separate enterprise features.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>8 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>api_definitions.go</strong><dd><code>Refactor UpstreamBasicAuth to use AuthSource struct</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apidef/api_definitions.go

<li>Changed <code>HeaderName</code> to <code>Header</code> in <code>UpstreamBasicAuth</code>.<br> <li> Introduced <code>AuthSource</code> struct for auth configurations.<br> <li> Added methods <code>IsEnabled</code> and <code>AuthKeyName</code> to <code>AuthSource</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6669/files#diff-9961ccc89a48d32db5b47ba3006315ef52f6e5007fb4b09f8c5d6d299c669d67">+24/-2</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>upstream.go</strong><dd><code>Update UpstreamBasicAuth to use AuthSource in OAS</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apidef/oas/upstream.go

<li>Changed <code>HeaderName</code> to <code>Header</code> in <code>UpstreamBasicAuth</code>.<br> <li> Updated <code>Fill</code> and <code>ExtractTo</code> methods to handle <code>AuthSource</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6669/files#diff-7b0941c7f37fe5a2a23047e0822a65519ca11c371660f36555b59a60f000e3f4">+18/-5</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>middleware.go</strong><dd><code>Implement Upstream Basic Auth Middleware</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ee/middleware/upstreambasicauth/middleware.go

<li>Added new middleware for upstream basic authentication.<br> <li> Implemented methods for middleware lifecycle and request processing.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6669/files#diff-4738eae6b1fa23f58598d492a9fa4fa6682545bf00bf122040180425a5e9dc60">+76/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>model.go</strong><dd><code>Define Middleware Model and APISpec Structures</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ee/middleware/upstreambasicauth/model.go

<li>Defined interfaces and structures for middleware dependencies.<br> <li> Added <code>APISpec</code> struct for middleware configuration.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6669/files#diff-94ad4a07e9dd6d82d74f8168df9c8f69f2635ab465cd2cce0e90def2846dae0a">+48/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>provider.go</strong><dd><code>Implement Provider for Upstream Authentication</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

ee/middleware/upstreambasicauth/provider.go

<li>Implemented <code>Provider</code> for upstream authentication.<br> <li> Added method to set request headers for authentication.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6669/files#diff-16e7043ed331108c72617c42eb50870a76523593a477bebb197fae04c22943a4">+26/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>api_loader.go</strong><dd><code>Integrate Upstream Basic Auth Middleware in API Loader</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

gateway/api_loader.go

- Integrated new upstream basic auth middleware into API processing.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6669/files#diff-cdf0b7f176c9d18e1a314b78ddefc2cb3a94b3de66f1f360174692c915734c68">+4/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>mw_upstream_basic_auth.go</strong><dd><code>Replace UpstreamBasicAuth with Noop for Non-EE Builds</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

gateway/mw_upstream_basic_auth.go

<li>Replaced <code>UpstreamBasicAuth</code> with <code>noopUpstreamBasicAuth</code>.<br> <li> Added build constraints for non-EE and non-dev environments.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6669/files#diff-ba603a8b249fdf72522258e825b7f9c64064203129c167795b206d66e9ebcda7">+14/-54</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>mw_upstream_basic_auth_ee.go</strong><dd><code>Add EE-Specific Upstream Basic Auth Middleware</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

gateway/mw_upstream_basic_auth_ee.go

<li>Added EE-specific implementation for upstream basic auth middleware.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6669/files#diff-bfe56451e8be136430a85c8c162d5d70135d04eac3097a911236ea24305089de">+14/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>mw_upstream_basic_auth_test.go</strong><dd><code>Update Test Build Constraints for EE Middleware</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

gateway/mw_upstream_basic_auth_test.go

- Added build constraints for EE and dev environments.

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6669/files#diff-15f78fac7fd4c8c0a1dcbd86ac6068e5a1a39f948f40afba6a6081e5f90f0ecd">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

**Comment by buger** [comment]
<p>:broken_heart: The detected issue is not in one of the allowed statuses :broken_heart: </p>    
   <table>
     <tr>
        <th>Detected Status</th>
        <td>Open</td>
        <td>:x:</td>
     </tr>
     <tr>
        <th>Allowed Statuses</th>
        <td>In Dev,In Code Review,Ready for Testing,In Test,In Progress,In Review</td>
        <td>:heavy_check_mark:</td>
      </tr>
   </table>
   <p>Please ensure your jira story is in one of the allowed statuses</p>

File: apidef/api_definitions.go [hunk 1]
```diff
@@ -797,9 +797,31 @@ type UpstreamBasicAuth struct {
 	Username string `bson:"username" json:"username"`
 	// Password is the password to be used for upstream basic authentication.
 	Password string `bson:"password" json:"password"`
-	// HeaderName is the custom header name to be used for upstream basic authentication.
+	// Header holds the configuration for custom header name to be used for upstream basic authentication.
```
Comment (titpetric) [inline apidef/api_definitions.go:800]: Not sure why this is changed?

File: ee/middleware/upstreambasicauth/model.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+// Gateway is the subset of Gateway APIs that the middleware uses.
+type Gateway interface {
+	model.ConfigProvider
+	model.ReplaceTykVariables
+}
+
+// APISpec is a subset of gateway.APISpec for the values the middleware consumes.
+type APISpec struct {
```
Comment (titpetric) [inline ee/middleware/upstreambasicauth/model.go:30]: I don't hate it, but why not make a complete copy of the struct in `model.`?
And ideally have an `func(*APISpec)Model()*model.APISpec` to bridge the gap with a single type.

That gets rid of a bunch of parameters in NewAPISpec or even delete it. Data models shouldn't have constructors, service models should and for what it's worth this is data model.

File: ee/middleware/upstreambasicauth/middleware.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+// Init initializes the middleware.
+func (m *Middleware) Init() {
+	m.Logger().Debug("Initializing Upstream basic auth Middleware")
+}
+
+// ProcessRequest will handle upstream basic auth.
+func (m *Middleware) ProcessRequest(_ http.ResponseWriter, r *http.Request, _ interface{}) (error, int) {
+	basicAuthConfig := m.Spec.UpstreamAuth.BasicAuth
+
+	upstreamBasicAuthProvider := Provider{
```
Comment (titpetric) [inline ee/middleware/upstreambasicauth/middleware.go:65]: This seems like a lot of code that could be a `NewProvider(m.Logger(), basicAuthConfig)`?

File: ee/middleware/upstreambasicauth/provider.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
+package upstreambasicauth
+
+import (
+	"net/http"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Provider implements upstream auth provider.
+type Provider struct {
```
Comment (titpetric) [inline ee/middleware/upstreambasicauth/provider.go:10]: This is a service struct which should have a constructor to pass the dependency (logger);
Bit weird to have it structured O(1) like this

File: apidef/api_definitions.go (hunk 1 above)
Comment (jeffy-mathew) [reply apidef/api_definitions.go:800]: as part of the contract change we had discussed yesterday - header should have capability to be enabled/disabled and can be extended to other auth sources

Comment (titpetric) [reply apidef/api_definitions.go:800]: not noted on ticket

File: gateway/mw_upstream_basic_auth_ee.go [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+package gateway
+
+import (
+	"github.com/TykTechnologies/tyk/ee/middleware/upstreambasicauth"
+)
+
+func getUpstreamBasicAuthMw(base *BaseMiddleware) TykMiddleware {
+	spec := base.Spec
+	mwSpec := upstreambasicauth.NewAPISpec(spec.APIID, spec.Name, spec.IsOAS, spec.OAS, spec.UpstreamAuth)
```
Comment (titpetric) [inline gateway/mw_upstream_basic_auth_ee.go:11]:
```suggestion
	mwSpec := spec.Model()
```

in line with a copy of the full APISpec in internal/model. Can use [huandu/go-clone](https://github.com/huandu/go-clone).

There might be a bridge already with model.MergedAPIs as apidef and oas objects are out of gateway scope. Meaning, you could take a model.MergedAPI today without the APISpec copy

File: ee/middleware/upstreambasicauth/model.go (hunk 2 above)
Comment (titpetric) [reply ee/middleware/upstreambasicauth/model.go:30]: APISpec.MergedAPI() part of follow up proposal in #6672, resolving

File: gateway/mw_upstream_basic_auth_ee.go (hunk 5 above)
Comment (titpetric) [reply gateway/mw_upstream_basic_auth_ee.go:11]: this is going away, resolving

**Review by titpetric — APPROVED** [review]
Resolved some improvements feedback, nothing open should be a blocker.

Follow up with #6672

**Pull request merged**